snuba/clickhouse/native.py

```diff
@@ -127,6 +131,67 @@ def close(self) -> None:
             pass
 
 
+T = TypeVar("T")
+R = TypeVar("R")
+
+
+def transform_nullable(
+    function: Callable[[T], R]
+) -> Callable[[Optional[T]], Optional[R]]:
+    def transform_column(value: Optional[T]) -> Optional[R]:
+        if value is None:
+            return value
+        else:
+            return function(value)
+
+    return transform_column
+
+
+def transform_date(value: date) -> str:
+    """
+    Convert a timezone-naive date object into an ISO 8601 formatted date and
+    time string respresentation.
+    """
+    # XXX: If the original value had a valid nonzero UTC offset, this will
+    # result in an incorrect value being returned.
```

Comment:
This is one of those "shouldn't happen" (but probably will eventually) comments. Mark my words.

Reply:
Did you figure out why some dates had timezone from clickhouse and others did not ? I am curious.
Also should we log warning in case the result we are returning is incorrect because there was a valid non zero UTC offset in the parameter ?

Reply:
> Did you figure out why some dates had timezone from clickhouse and others did not ? I am curious.

It seems like certain functions that manipulate the time value result in the time having the timezone attached, as in these examples:

```sql
SELECT now() LIMIT 1 FORMAT JSON
```
```json
{
	"meta":
	[
		{
			"name": "now()",
			"type": "DateTime"
		}
	],
	"data":
	[
		{
			"now()": "2020-03-04 22:57:42"
		}
	]
}
```

vs.

```sql
SELECT toStartOfHour(now()) LIMIT 1 FORMAT JSON
```
```json
{
	"meta":
	[
		{
			"name": "toStartOfHour(now())",
			"type": "DateTime('Etc\/UTC')"
		}
	],
	"data":
	[
		{
			"toStartOfHour(now())": "2020-03-04 22:00:00"
		}
	]
}
```

In our case, this means there is going to be a difference between the values returned by the `time` synthetic column added by the time series dataset code, and the `timestamp` returned from a single row.

> Also should we log warning in case the result we are returning is incorrect because there was a valid non zero UTC offset in the parameter ?

I thought about this too. I don't like that this will be silently adjusting (read: corrupting) data, but also the idea of logging within the value transformer seems like a bad idea since it'll put the log in a tight loop. This also seems to be inconsistently named between environments (in production it is `Universal` whereas I see `Etc/UTC`.

I think that the real answer is to just treat tz-naive and tz-aware datetimes differently. If there's no tzinfo, assume UTC and if there is tzinfo, either ensure (via asserting) that it's UTC, or always localize it to UTC regardless.

Reply:
> I think that the real answer is to just treat tz-naive and tz-aware datetimes differently. If there's no tzinfo, assume UTC and if there is tzinfo, either ensure (via asserting) that it's UTC, or always localize it to UTC regardless.

I'm going to do this as a follow up just to do some semi-dangerous thing at a time.